File: leetcode/039-combination-sum.cpp
class Solution
{
 public:
  vector<vector<int> >
  combinationSum(vector<int>& candidates, int target)
  {
    dfs(candidates, 0, target);
    return ret_;
  }

  void dfs(vector<int>& cands, int i0, int target)
  {
    if (0 > target) return;
    if (0 == target) {
      ret_.push_back(sol_);
      return;
    }

    sol_.push_back(0);
    for (int i = i0; i < cands.size(); ++i) {
      sol_.back() = cands[i];
      dfs(cands, i, target - cands[i]);
    }
    sol_.pop_back();
  }

 private:
  vector<int> sol_;
  vector<vector<int> > ret_;
};

File: leetcode/039-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        def dfs(cands, i0, target, ret, sol):
            if 0 > target:
                return
            if 0 == target:
                ret += [sol[:]]
                return
            for i in range(i0, len(cands)):
                dfs(cands, i, target-cands[i], ret, sol+[cands[i]])

        ret = []
        dfs(candidates, 0, target, ret, [])
        return ret
